src/stream_base.h

```diff
@@ -146,10 +146,14 @@ class StreamResource {
                          const uv_buf_t* buf,
                          uv_handle_type pending,
                          void* ctx);
+  typedef void (*DestructCb)(void* ctx);
 
   StreamResource() : bytes_read_(0) {
   }
-  virtual ~StreamResource() = default;
+  virtual ~StreamResource() {
+    if (!destruct_cb_.is_empty())
+      destruct_cb_.fn(destruct_cb_.ctx);
```

Comment:
@bnoordhuis setting `stream_ = nullptr` in `~TLSWrap()` actually isn't enough. And actually, just managed to create a new test that still segfaults. So I'm revisiting the patch.